### ISSUE TITLE
Possible memory leak on StreamProvider.autoDispose

### ISSUE THREAD

**Issue by fpabl0** [rrousselGit/river_pod#193]
**Describe the bug**
I am trying to replace StreamBuilders in favor of StreamProvider.autoDispose, however they don't behave equally (don't know if I am misunderstanding StreamProvider, if so, then sorry). Using watch(streamProvider) does not unsubscribe from the stream when the ui is destroyed as a consequence, every time I returned to that screen a new subscription is done with disposing the previous one. That's not the case with the StreamBuilder.

**To Reproduce**

I have this method in a class that return a stream:

```dart
final waypointsRepoProvider = Provider<WaypointsRepository>(
  (ref) => WaypointsFirestoreRepository(),
);

class WaypointsFirestoreRepository implements WaypointsRepository {
  @override
  Stream<List<WaypointModel>> getWaypointsStream() {
    final route = FirebaseFirestore.instance
            .collection("drivers")
            .doc("J5AIHrG0CH3kOpJUpwCq")
            .collection("route");
    return route.snapshots().map<List<WaypointModel>>((snapshot) {
      print("here"); // this message is the key to see if the stream has still listeners when the ui is destroyed
      return snapshot.docs.map((el) {
        return WaypointModel(
          position: el.get("position"),
          address: el.get("address"),
          completed: el.get("completed"),
        );
      }).toList();
    });
  }
}
```
With StreamBuilder (when user goes to other screen, if data changes "here" is not displayed, because the stream subscription has been closed by the streambuilder):
```dart
  Widget build(BuildContext context, ScopedReader watch) {
    final repo = watch(waypointsRepoProvider);
    return StreamBuilder(
      stream: repo.getWaypointsStream(),
      builder: (context, snapshot) => Container(
        child: snapshot.data != null
            ? Text("${snapshot.data[0]?.address}")
            : Text("no data"),
      ),
    );
 }
```
With StreamProvider (when user goes to other screen, "here" message still triggers when data changes, and you can get a lot of them at once, if you returned to the ui that use the stream many times):
```dart
final waypointsStreamProvider = StreamProvider.autoDispose<List<WaypointModel>>((ref) {
  final repo = ref.watch(waypointsRepoProvider);
  return repo.getWaypointsStream();
});
```
```dart
  Widget build(BuildContext context, ScopedReader watch) {
    final waypoints = watch(waypointsStreamProvider);
    return waypoints.maybeWhen(
      data: (w) => Container(child: Text("${w[0].address}")),
      orElse: () => Container(child: Text("no data")),
    );
 }
```
**Expected behavior**
I expect that the subscriptions are totally disposed when the ui that use the streamProvider is destroyed.

**Comment by rrousselGit** [comment]
Could you share fully executable examples with step by steps way to reproduce the problem?

**Comment by fpabl0** [comment]
Yes, of course, I have created a github repository with an example with the problem:
https://github.com/fpabl0/riverpod_streamprovider_test

**Comment by rrousselGit** [comment]
Thanks!

I'm able to reproduce the issue. I will investigate

**Comment by rrousselGit** [comment]
I think that's a bug related to broadcast streams:

```dart
void main() {
  final stream = Stream.periodic(Duration(seconds: 5), (n) {
    print('$n');
    return n;
  });

  final sub = stream.asBroadcastStream().listen((value) {
    print('value $value');
  });

  print('cancel');
  sub.cancel();
}
```

This will print `cancel` then keep printing an incrementing number (but the "value" print does not appear)

**Comment by rrousselGit** [comment]
Interesting case. thanks for the report! Learned something new today

TL;DR, the bug was, Riverpod does:

```dart
StreamSubscription sub;
Stream publicStream;

initState() {
  publicStream = stream.asBroadcastStream();
  sub = publicStream.listen(...);
}

dispose() {
  sub.cancel();
}
```


But with this, while the subscriptions are closed, the broadcast stream never closes

The fix appears to be:

```dart
StreamSubscription sub;
Stream publicStream;
StreamController controller;

initState() {
  controller = StreamController.broadcast();
  publicStream = controller.stream;
  sub = publicStream.listen(controller.add, onError: controller.addError, onDone: controller.close);
}

dispose() {
  sub.cancel();
  controller.close();
}
```

**Comment by fpabl0** [comment]
> ```dart
> void main() {
>   final stream = Stream.periodic(Duration(seconds: 5), (n) {
>     print('$n');
>     return n;
>   });
> 
>   final sub = stream.asBroadcastStream().listen((value) {
>     print('value $value');
>   });
> 
>   print('cancel');
>   sub.cancel();
> }
> ```

Hmm but in that case the value printed is because of the stream generator, the case in the example is this:
```dart
void main() async {
  final stream = Stream.periodic(Duration(seconds: 1), (n) {
    print('$n');
    return n;
  });

  final sub = stream.asBroadcastStream().map((event) {
    print("here");
    return event;
  }).listen((value) {
    print('value $value');
  });

  print('cancel');
  sub.cancel();

  await Future.delayed(Duration(seconds: 5));
}
```
You will see the numbers printed, but not "here" message, because the subscription was cancelled.

**Comment by fpabl0** [comment]
> Interesting case. thanks for the report! Learned something new today
> 
> TL;DR, the bug was, Riverpod does:
> 
> ```dart
> StreamSubscription sub;
> Stream publicStream;
> 
> initState() {
>   publicStream = stream.asBroadcastStream();
>   sub = publicStream.listen(...);
> }
> 
> dispose() {
>   sub.cancel();
> }
> ```
> 
> But with this, while the subscriptions are closed, the broadcast stream never closes
> 
> The fix appears to be:
> 
> ```dart
> StreamSubscription sub;
> Stream publicStream;
> StreamController controller;
> 
> initState() {
>   controller = StreamController.broadcast();
>   publicStream = controller.stream;
>   sub = publicStream.listen(controller.add, onError: controller.addError, onDone: controller.close);
> }
> 
> dispose() {
>   sub.cancel();
>   controller.close();
> }
> ```

Thanks to you for this amazing library. 
Hmm yes, I didn't know StreamProvider creates internally a broadcast stream, that should be the solution,

**Comment by rrousselGit** [comment]
Yes but that is not what happen here.

The code is not `stream.asBroadcast().map()` but `stream.map().asBroadcast`

**Comment by fpabl0** [comment]
> Yes but that is not what happen here.
> 
> The code is not `stream.asBroadcast().map()` but `stream.map().asBroadcast`

Yep, I see, I didn't realize StreamProvider creates internally a broadcast stream.

**Comment by rrousselGit** [comment]
Yeah

The broadcast stream is needed for `watch(streamProvider.stream).listen(...)`

**Comment by fpabl0** [comment]
> Yeah
> 
> 
> 
> The broadcast stream is needed for `watch(streamProvider.stream).listen(...)`

Oh, I see, so that should be the solution, closing the StreamController in the dispose method.

**Comment by fpabl0** [comment]
Hi @rrousselGit , sorry for comment on this closed issue, I just want to share that I have just realized that the broadcast stream can be closed using asBroadcastStream method too:
```dart
void main() async {
  final stream = Stream.periodic(Duration(seconds: 1), (n) {
    print('$n');
    return n;
  });

  final broadcastStream = stream.asBroadcastStream(
    onCancel: (sub) {
      print("don't have listeners now, so close the underlying subscription");
      sub.cancel();
    },
  );

  final sub1 = broadcastStream.listen((value) {
    print('value sub1 -> $value');
  });

  final sub2 = broadcastStream.listen((value) {
    print('value sub2 -> $value');
  });

  print('cancelling sub1');
  sub1.cancel();
  print('cancelling sub2');
  sub2.cancel();

  await Future.delayed(Duration(seconds: 5));
}
```

**Comment by rrousselGit** [comment]
Indeed

But in this case it makes more sense to have a StreamController

**Comment by fpabl0** [comment]
@rrousselGit version 0.12.0 solves this issue?

**Comment by rrousselGit** [comment]
Yes it does

**Comment by fpabl0** [comment]
> Yes it does

Hmm I have updated the example to that version, however I am still getting the same problem.

**Comment by rrousselGit** [comment]
Ah my bad, I fixed it for `StreamProvider` but not `autoDispose` (which had the issue with `context.refresh`/`ref.watch` too)

**Comment by rrousselGit** [comment]
I'll deploy another fix later today, it's a small change

**Comment by fpabl0** [comment]
> I'll deploy another fix later today, it's a small change

Perfect, thank you so much 😄

**Comment by muhajirdev** [comment]
Hi @rrousselGit @fpabl0 ,

```dart
final userProvider = StreamProvider((ref) {
  return FirebaseAuth.instance.authStateChanges();
});

final firebaseClassData = StreamProvider.autoDispose((ref) {
  final userStream = ref.watch(userProvider.last);

  final result = userStream.asStream().switchMap((user) {
    print('got user $user');
    return FirebaseFirestore.instance
        .collection('users')
        .doc(user.uid)
        .collection('courses')
        .snapshots()
        .map((snapshots) => snapshots.docs);
  }).map((event) {
    print('snapshot is $event');
    return event;
  });

  print('result is $result');

  return result;
});

final classListProvider =
    FutureProvider.autoDispose<List<ClassData>>((ref) async {
  final prismicClass = await ref.watch(allClassesProvider.future);
  print('got prismicClass data $prismicClass');
  final userClassData = await ref.watch(firebaseClassData.last);

  final result = prismicClass.results
      .map((result) => getClassData(result, userClassData))
      .toList();

  return result;
});

```

So my code looks like this. For some reason if I see the `debug console` on my vscode. It seems like firebaseClassData keeps re running again and again. 

The logs looks like this

```
flutter: snapshot is [Instance of 'QueryDocumentSnapshot']
flutter: got prismicClass data Instance of 'PrismicClassData'
flutter: result is Instance of '_MapStream<List<QueryDocumentSnapshot>, List<QueryDocumentSnapshot>>'
flutter: got user User(displayName: null, email: xxx@gmail.com, emailVerified: true, isAnonymous: false, metadata: UserMetadata(creationTime: 2020-12-07 12:34:41.919, lastSignInTime: 2020-12-11 18:02:38.788), phoneNumber: null, photoURL: null, providerData, [UserInfo(displayName: Muhammad Muhajir, email: xxx@gmail.com, phoneNumber: null, photoURL: https://lh4.googleusercontent.com/-4wjF-SdZsNw/AAAAAAAAAAI/AAAAAAAABkI/s96-c/photo.jpg, providerId: google.com, uid: 116894881093999019464), UserInfo(displayName: null, email: xxx@gmail.com, phoneNumber: null, photoURL: null, providerId: apple.com, uid: 000306.cfd3ad86aa7c4344a0d5930ea8fe6999.1019)], refreshToken: , tenantId: null,  hZPCJNyk93aR1AYoAwimTyKVl9c2)
flutter: snapshot is [Instance of 'QueryDocumentSnapshot']
flutter: got prismicClass data Instance of 'PrismicClassData'
flutter: result is Instance of '_MapStream<List<QueryDocumentSnapshot>, List<QueryDocumentSnapshot>>'
flutter: got user User(displayName: null: xxx@gmail.com, emailVerified: true, isAnonymous: false, metadata: UserMetadata(creationTime: 2020-12-07 12:34:41.919, lastSignInTime: 2020-12-11 18:02:38.788), phoneNumber: null, photoURL: null, providerData, [UserInfo(displayName: Muhammad Muhajir, email: xxx@gmail.com, phoneNumber: null, photoURL: https://lh4.googleusercontent.com/-4wjF-SdZsNw/AAAAAAAAAAI/AAAAAAAABkI/s96-c/photo.jpg, providerId: google.com, uid: 116894881093999019464), UserInfo(displayName: null, email: xxx@gmail.com, phoneNumber: null, photoURL: null, providerId: apple.com, uid: 000306.cfd3ad86aa7c4344a0d5930ea8fe6999.1019)], refreshToken: , tenantId: null,  hZPCJNyk93aR1AYoAwimTyKVl9c2)
flutter: snapshot is [Instance of 'QueryDocumentSnapshot']
flutter: got prismicClass data Instance of 'PrismicClassData'
flutter: result is Instance of '_MapStream<List<QueryDocumentSnapshot>, List<QueryDocumentSnapshot>>'
flutter: got user User(displayName: null: xxx@gmail.com, emailVerified: true, isAnonymous: false, metadata: UserMetadata(creationTime: 2020-12-07 12:34:41.919, lastSignInTime: 2020-12-11 18:02:38.788), phoneNumber: null, photoURL: null, providerData, [UserInfo(displayName: Muhammad Muhajir, email: xxx@gmail.com, phoneNumber: null, photoURL: https://lh4.googleusercontent.com/-4wjF-SdZsNw/AAAAAAAAAAI/AAAAAAAABkI/s96-c/photo.jpg, providerId: google.com, uid: 116894881093999019464), UserInfo(displayName: null, email: xxx@gmail.com, phoneNumber: null, photoURL: null, providerId: apple.com, uid: 000306.cfd3ad86aa7c4344a0d5930ea8fe6999.1019)], refreshToken: , tenantId: null,  hZPCJNyk93aR1AYoAwimTyKVl9c2)
flutter: snapshot is [Instance of 'QueryDocumentSnapshot
```


But, if I change `final firebaseClassData = StreamProvider.autoDispose((ref) {` into `final firebaseClassData = StreamProvider((ref) {`

(removing `autoDispose`). Things seems to be working okay. Is this is a bug or I missed something ?

**Comment by rrousselGit** [comment]
It is likely that for a few frames, your `firebaseClassData` is no-longer listened and therefore destroyed

**Comment by muhajirdev** [comment]
Hi @rrousselGit . Thanks for a the quick reply

So I tried to reproduce this issue without firebase

```
final testProvider = StreamProvider.autoDispose((ref) async* {
  ref.onDispose(() {
    print('testProvider disposed');
  });

  yield '1';
  yield '2';
  yield '3';
  yield '4';
});

final someFutureProvider = FutureProvider((ref) async {
  print('future is called');
  return 'something';
});

final test2Provider = FutureProvider.autoDispose((ref) async {
  final somefuture = await ref.watch(someFutureProvider.future);
  print('somefuture is $somefuture');
  final a = await ref.watch(testProvider.last);
  print('a is $a');

  return 'c';
});

class Home extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Scaffold(body: Center(
      child: Consumer(
        builder: (context, watch, child) {
          final u = watch(test2Provider);
          return u.when(
              data: (a) => Text(
                    "anything",
                    style: TextStyle(color: Colors.white),
                  ),
              loading: () => Text(
                    'a',
                    style: TextStyle(color: Colors.white),
                  ),
              error: (e, s) => Text(
                    'b',
                    style: TextStyle(color: Colors.white),
                  ));
        },
      ),
    ));


  }}
```

try to switch

```
 final somefuture = await ref.watch(someFutureProvider.future);
  print('somefuture is $somefuture');
  final a = await ref.watch(testProvider.last);
  print('a is $a');
```

with
```
final a = await ref.watch(testProvider.last);
print('a is $a');
final somefuture = await ref.watch(someFutureProvider.future);
print('somefuture is $somefuture');
```

what I found is, if I watch futureProvider first, then I'l' got infinite log like this

```
flutter: testProvider disposed
flutter: somefuture is something
flutter: a is 1
flutter: testProvider disposed
flutter: somefuture is something
flutter: a is 1
flutter: testProvider disposed
flutter: somefuture is something
flutter: a is 1
flutter: testProvider disposed
flutter: somefuture is something
flutter: a is 1
flutter: testProvider disposed
flutter: somefuture is something
flutter: a is 1
flutter: testProvider disposed
flutter: somefuture is something
flutter: a is 1
flutter: testProvider disposed
flutter: somefuture is something
flutter: a is 1
flutter: testProvider disposed
flutter: somefuture is something
flutter: a is 1
flutter: testProvider disposed
flutter: somefuture is something
flutter: a is 1
flutter: testProvider disposed
flutter: somefuture is something
flutter: a is 1
flutter: testProvider disposed
```

But if I switch the order, it will be okay

**Comment by muhajirdev** [comment]
It looks like the StreamProvider is dispossed and then recreated

**Comment by muhajirdev** [comment]
Btw what I am trying to accomplish here is to combine data from two sources. From firestore (stream) and from an API (future).

Am i doing it the right way?

**Comment by rrousselGit** [comment]
It may be a bug, I'll look at it later. Can you create a new issue?

**Comment by muhajirdev** [comment]
I opened new issue @rrousselGit
